File: frontend/src/api/identities.ts
import { client } from './client'

// ── Types ─────────────────────────────────────────────────────────────────────

export type MergeRequest = {
  account_ids: string[]
}

export type MergeResponse = {
  identity_id: string
  account_ids: string[]
}

export type UnmergeRequest = {
  account_id: string
}

export type UnmergeResponse = {
  caveat: string
}

export type SuggestionItem = {
  account_a_id: string
  account_b_id: string
  score: number
  reasons: string[]
}

export type SuggestionsResponse = {
  suggestions: SuggestionItem[]
  /** Always false until Phase 5 (txn dedup) ships — do NOT render aggregated balances. */
  aggregated: false
}

export type DismissSuggestionRequest = {
  account_a_id: string
  account_b_id: string
  write_do_not_merge?: boolean
}

export type DismissSuggestionResponse = {
  dismissed: boolean
}

export type IdentityMember = {
  id: string
  user_id: string
  provider_account_id: string
  account_type: string
  currency: string
  display_name: string | null
  identity_id: string | null
  created_at: string
}

export type IdentityResponse = {
  id: string
  user_id: string
  origin: string
  master_account_id: string | null
  merged_into: string | null
  merged_at: string | null
  created_at: string
  updated_at: string
  members: IdentityMember[]
  /** Always false until Phase 5 (txn dedup) ships — do NOT render aggregated balances. */
  aggregated: false
}

// ── API functions ─────────────────────────────────────────────────────────────

export const identitiesApi = {
  async merge(payload: MergeRequest) {
    const response = await client.post<MergeResponse>('/api/v1/identities/merge', payload)
    return response.data
  },

  async unmerge(payload: UnmergeRequest) {
    const response = await client.post<UnmergeResponse>('/api/v1/identities/unmerge', payload)
    return response.data
  },

  async suggestions() {
    const response = await client.get<SuggestionsResponse>('/api/v1/identities/suggestions')
    return response.data
  },

  async dismissSuggestion(payload: DismissSuggestionRequest) {
    const response = await client.post<DismissSuggestionResponse>(
      '/api/v1/identities/suggestions/dismiss',
      payload,
    )
    return response.data
  },

  async getIdentity(identityId: string) {
    const response = await client.get<IdentityResponse>(`/api/v1/identities/${identityId}`)
    return response.data
  },
}

File: backend/src/manifold/schemas/identities.py
from __future__ import annotations

from datetime import datetime

from pydantic import field_validator

from manifold.schemas.common import SchemaModel


class MergeRequest(SchemaModel):
    """Request body for POST /identities/merge."""

    account_ids: list[str]

    @field_validator("account_ids")
    @classmethod
    def at_least_two(cls, v: list[str]) -> list[str]:
        if len(v) < 2:
            raise ValueError("account_ids must contain at least 2 entries")
        return v


class MergeResponse(SchemaModel):
    """Response for POST /identities/merge."""

    identity_id: str
    account_ids: list[str]


class UnmergeRequest(SchemaModel):
    """Request body for POST /identities/unmerge."""

    account_id: str


class UnmergeResponse(SchemaModel):
    """Response for POST /identities/unmerge."""

    caveat: str


class SuggestionItem(SchemaModel):
    """One merge suggestion pair."""

    account_a_id: str
    account_b_id: str
    score: float
    reasons: list[str]


class SuggestionsResponse(SchemaModel):
    """Response for GET /identities/suggestions."""

    suggestions: list[SuggestionItem]
    aggregated: bool = False


class DismissSuggestionRequest(SchemaModel):
    """Request body for POST /identities/suggestions/dismiss."""

    account_a_id: str
    account_b_id: str
    write_do_not_merge: bool = True


class DismissSuggestionResponse(SchemaModel):
    """Response for POST /identities/suggestions/dismiss."""

    dismissed: bool


class IdentityMemberResponse(SchemaModel):
    """One account member of an identity."""

    id: str
    user_id: str
    provider_account_id: str
    account_type: str
    currency: str
    display_name: str | None
    identity_id: str | None
    created_at: datetime


class IdentityResponse(SchemaModel):
    """Response for GET /identities/{id}."""

    id: str
    user_id: str
    origin: str
    master_account_id: str | None
    merged_into: str | None
    merged_at: datetime | None
    created_at: datetime
    updated_at: datetime
    members: list[IdentityMemberResponse]
    aggregated: bool = False


__all__ = [
    "DismissSuggestionRequest",
    "DismissSuggestionResponse",
    "IdentityMemberResponse",
    "IdentityResponse",
    "MergeRequest",
    "MergeResponse",
    "SuggestionItem",
    "SuggestionsResponse",
    "UnmergeRequest",
    "UnmergeResponse",
]
